### ISSUE TITLE
Support the immediate release of translated chapters

### ISSUE THREAD

**Issue by rviscomi** [HTTPArchive/almanac.httparchive.org#671]
We've made amazing progress translating the contents of the Almanac into multiple languages. Unfortunately, none of this work is publicly available because the translations are being batched up to be released when the entire contents are ready.

What needs to happen for us to start releasing translated chapters as soon as they're completed?

cc @HTTPArchive/developers @HTTPArchive/translators 

--------

Edit by @bazzadp , listing tasks here:

- [x] Add ability to mark pages (chapters or the non-chapter pages) as available in other languages and if so which languages. Maybe in a similar manner as to how we did the "unedited" note on each chapter. That was originally done as "todo" in the JSON config, but then we moved to unedited in the chapter meta-data but think this would be better stored in the JSON config as otherwise would need to store each variation that's available multiple times in each chapter.
- [x] Update the translation drop down (currently hidden) to only display when translation exists (as specified in previous point).
- [x] Remove top "Start Exploring" link as discussed in #574 (comment) as not enough space for it once we add the Language switcher (and some of us think it's just wrong anyway 😀)
- [x] Review Language switcher in all screen sizes and browsers to ensure it is in correct place as not really been tested much and templates have changed a fair bit since it was done.
Add missing templates for the other Languages were we have translations. Only the French translation has all the templates needed to be able to generate the HTML for the chapters so only it can launch and the Spanish and Japanese versions can't. To be honest I'd prefer to work on #673 first to hopefully make this easier for the Spanish and Japanese translators.
- [x] Fix #654 before we can launch some of the French chapters.
- [ ] Decide what we are doing with #567 - maybe live with it as is for initial launch?
- [x] Fix #542 and #625 before we can launch the Spanish chapters.
- [ ] I gather visualisations (#558) are out of scope for now and we circle back to them once the text has been translated if there's still volunteers to tackle that?
- [x] Redirect any missing translations back to English so any references to them in translated chapters don't 404. This should be a 302 (temporary redirect) instead of a 301 (permanent redirect) for SEO purposes. We should also check the cache time of this.

**Comment by bazzadp** [comment]
Yeah we definitely should look to do this. It's a shame to hold back all the good work fo the translations rather than release them as we can (as Wikipedia and MDN do!). Will also hopefully encourage more translations too as it's been slow progress and I'd imagine not very encouraging when you won't see the fruits of your labour until the end.

Some things I think we'd need to do to allow this to happen:

- Add ability to mark pages (chapters or the non-chapter pages) as available in other languages and if so which languages. Maybe in a similar manner as to how we did the "unedited" note on each chapter. That was [originally done as "todo" in the JSON config, but then we moved to unedited in the chapter meta-data](https://github.com/HTTPArchive/almanac.httparchive.org/commit/7769b95aac758c0c84dd6268334e15ecffe0b35e) but think this would be better stored in the JSON config as otherwise would need to store each variation that's available multiple times in each chapter.
- Update the translation drop down (currently hidden) to only display when translation exists (as specified in previous point).
- Remove top "Start Exploring" link as discussed in https://github.com/HTTPArchive/almanac.httparchive.org/issues/574#issuecomment-571364742 as not enough space for it once we add the Language switcher (and some of us think it's just wrong anyway 😀)
- Review Language switcher in all screen sizes and browsers to ensure it is in correct place as not really been tested much and templates have changed a fair bit since it was done.
- Add missing templates for the other Languages were we have translations. Only the French translation has all the templates needed to be able to generate the HTML for the chapters so only it can launch and the Spanish and Japanese versions can't. To be honest I'd prefer to work on #673 first to hopefully make this easier for the Spanish and Japanese translators.
- Fix #654 before we can launch some of the French chapters.
- Decide what we are doing with #567 - maybe live with it as is for initial launch?
- Fix #542 and #625 before we can launch the Spanish chapters.
- I gather visualisations (#558) are out of scope for now and we circle back to them once the text has been translated if there's still volunteers to tackle that?
- Redirect any missing translations back to English so any references to them in translated chapters don't 404. This should be a 302 (temporary redirect) instead of a 301 (permanent redirect) for SEO purposes. We should also check the cache time of this.

Thats all I can think off now and, if you agree then we should probably add them as check boxes to first comment so we can tick them off as we do them.

**Comment by rviscomi** [comment]
> Add ability to mark pages (chapters or the non-chapter pages) as available in other languages and if so which languages. Maybe in a similar manner as to how we did the "unedited" note on each chapter. That was originally done as "todo" in the JSON config, but then we moved to unedited in the chapter meta-data but think this would be better stored in the JSON config as otherwise would need to store each variation that's available multiple times in each chapter.

Let's say only the JS chapter is translated to a particular language. A user lands on the translated JS chapter and tries to continue on to the CSS chapter. If the CSS chapter is only available in English, does the link take them to <lang>/2019/css or en/2019/css? It's not great but I think the first option works. So for a language to be supported, it doesn't necessarily need all content translated it just needs to have all of the templates/chapters copied over to its <lang> directory. This might actually be good for the translators because the PRs would show a diff of English to the translated version. It also simplifies our routing logic so once you land in <lang> you stay there until you explicitly change languages. Not sure what it does to SEO. This could be a good use of a flag similar to the "unedited" one, where we mark a chapter as _not translated yet_. That could trigger a little message on the page to help us translate it on GitHub.

> Update the translation drop down (currently hidden) to only display when translation exists (as specified in previous point).

If we go the route described above, for any language that has some translations done, we would add it to the list of available languages in the dropdown.

> Remove top "Start Exploring" link as discussed in #574 (comment) as not enough space for it once we add the Language switcher (and some of us think it's just wrong anyway 😀)

SGTM

> Review Language switcher in all screen sizes and browsers to ensure it is in correct place as not really been tested much and templates have changed a fair bit since it was done.

+1

> Add missing templates for the other Languages were we have translations. Only the French translation has all the templates needed to be able to generate the HTML for the chapters so only it can launch and the Spanish and Japanese versions can't. To be honest I'd prefer to work on #673 first to hopefully make this easier for the Spanish and Japanese translators.

I don't think #673 is blocking. Let's copy the English templates over so we can start serving _some_ translated content sooner.

> Fix #654 before we can launch some of the French chapters.

+1 we need to know what we're doing with the URL slugs. I think it's ok to keep the English naming in the URL? WDYT?

> Decide what we are doing with #567 - maybe live with it as is for initial launch?

I think it's ok to leave it in English. If absolutely needed we could regenerate the logo in HTML/CSS or make new images for each language. Not blocking.

> Fix #542 and #625 before we can launch the Spanish chapters.

SGTM

> I gather visualisations (#558) are out of scope for now and we circle back to them once the text has been translated if there's still volunteers to tackle that?

SGTM

> Redirect any missing translations back to English so any references to them in translated chapters don't 404. This should be a 302 (temporary redirect) instead of a 301 (permanent redirect) for SEO purposes. We should also check the cache time of this.

Maybe it's ok to serve the English version from the <lang> URL? If we show a translated note on the page that "this chapter isn't translated yet. if you can help, join the translation effort on github..." maybe that's ok. And if we're worried about SEO I think we could explicitly set noindex on untranslated pages.

**Comment by bazzadp** [comment]
> Let's say only the JS chapter is translated to a particular language. A user lands on the translated JS chapter and tries to continue on to the CSS chapter. If the CSS chapter is only available in English, does the link take them to `<lang>`/2019/css or en/2019/css?

> Maybe it's ok to serve the English version from the `<lang>` URL? If we show a translated note on the page that "this chapter isn't translated yet. if you can help, join the translation effort on github..." maybe that's ok. And if we're worried about SEO I think we could explicitly set noindex on untranslated pages.

I'm less of a fan of including the english content under `<lang>`/2019/css with a "translation required" flag/notice to be honest. I can see why the advantages (encourages translation, keeps them in lang), but I've just never seen that done before in other sites to be honest so seems a little weird to me. And for some languages they could be up a long time.

Definitely would need a `noindex` to avoid SEO issues if we did go down this route.

But also what language to say in `<html>` tag? `<html lang="fr">` when content in English is just wrong and `<html lang="en">` for `fr` URL seems weird (though more correct based on content). Just one of many things to consider.

Plus we're false advertising on other pages (you get sent English link but notice the drop down and think we have this page available in your local language - cool! click. dissappointment 😢). Unless you're saying we still only show the translated languages in the drop downs if an actual translation exists and the `<lang>`/2019/css is only given if explicitly navigated to (e.g. from `<lang>`/2019/javascript)? But then we're not making the fact we have some chapters in that language as explicit as you could so going against point of having English ones?

All in all I just think it's more pain than it's worth and very unusual to have English placeholders. If it was an ultra short term thing (like the unedited) I could be persuaded but don't think it will be for many languages.

> I don't think #673 is blocking. Let's copy the English templates over so we can start serving some translated content sooner.

Mmmm... I'd really like to get this out to be honest. Risk of someone updating an English Language template and forgetting one or more other languages is quite high IMHO (some of the are already our of date) and impact would be higher after launching a language (as we could breaking a production page if we forgot to update a non-English language template) but possibly more difficult to see (can we really hand on heart say we'd test all languages for every release, and even if we did, spot all problems?).

> we need to know what we're doing with the URL slugs. I think it's ok to keep the English naming in the URL? WDYT?

Absolutely OK in my book and very common (see [MDN](https://developer.mozilla.org/es/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images) or [Wikipedia](https://de.wikipedia.org/wiki/Responsive_Webdesign) - though Wikipedia do also sometimes translate URLs ([for example](https://fr.wikipedia.org/wiki/Site_web_r%C3%A9actif)). I think we could add [redirects](https://github.com/HTTPArchive/almanac.httparchive.org/blob/9801d8bf4f72e937304c78280e9fb0baf5aad518/src/validate.py#L42) for languages for convience but have the canonical in English even for other languages.

On that point do we see many 404s that we should add other redirects for? Could you permission me to Google Search Console so I can have a look and maybe keep an eye on this?

**Comment by bazzadp** [comment]
Here’s a, perhaps crazy, idea: what about waiting until we have a minimum number of chapters translated (E.g. 5 of them) and then machine translating the rest of them (using Chrome/Google Translate) in the short term to allow launching that language?

Could then add a “Machine translated - please get in touch if you want to help translate this properly” message.

I think that would be better than showing the English version under the translated URL.

**Comment by borisschapira** [comment]
Hi! There's two things, here: what's best and what's quicker and more convenient to set up.

I think that adding templates should be the first step for every translation project, and translation teams should be encouraged to do so. I would have added a template to localize a message saying 

```
This page is not translated yet. If you would like to participate in its translation into {lang}, feel free to <a href="https://github.com/HTTPArchive/almanac.httparchive.org/issues/{issue}" title="Help translating this page into {lang}">join the discussion</a>."
```

Then, we can add the chapter in English to every translation folder with a Front-Matter param containing the Id of the the translation issue. If the param is set, the abovementioned piece of text is displayed, the html tag displays an "en" `lang` attribute, and the page is not indexed. If the param is not set, the page is fully translated.

In the language selector, indeed, we can add an "add a translation" option, then redirect people to… [this research](https://github.com/HTTPArchive/almanac.httparchive.org/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+%28translate+OR+translation%29)?

I think these changes would be pretty fast to implement, while being satisfactory enough.

The only negative side of this approach is that the URLs would not be located. I do not realize the importance of the problem, but I think that, at worst, if we later realize that it is inconvenient, we will be able to implement a fix without invalidating the work previously done.

**Comment by bazzadp** [comment]
> I think that adding templates should be the first step for every translation project, and translation teams should be encouraged to do so.

I kind of agree. However translating the templates did require a fair bit of knowledge of the build system and which parts of the template to translate and not. We also didn't want to put people off translating chapters by demanding this knowldge.

Also it meant an awful lot of keeping the templates up to date. For example the Arabian chaper has a template (to [solve the right-to-left issue](https://github.com/HTTPArchive/almanac.httparchive.org/pull/551)) but then didn't make any progress on that language (yet) so we've just given ourselves work to maintain that template for no real reason yet. Which, as someone who's done quite a bit of work on large refactoring PRs for accessiblity and the like proved quite painful. Also meant this template was first alphabetically so comments were put on that for PR reviews. Perhaps that Arabic template should be removed again for now but then might forget how to do RTL in the templates.

Once #676 lands (assuming we're happy with this) then this all becomes **a lot** easier and think we can insist on that as part of each language and as part of the first PR for any new language added. This is why I really think we should land that before progressing further on this. Have a look at the [French base file](https://github.com/HTTPArchive/almanac.httparchive.org/blob/refactor_templates/src/templates/fr/2019/base.html) and the [French home page file](https://github.com/HTTPArchive/almanac.httparchive.org/blob/refactor_templates/src/templates/fr/2019/index.html) to see how that is set up in that branch. On on that note any additional reviews of #676 would be much appreciate! 😉

> I would have added a template to localize a message saying
> 
> ```
> This page is not translated yet. If you would like to participate in its translation into {lang}, feel free to <a href="https://github.com/HTTPArchive/almanac.httparchive.org/issues/{issue}" title="Help translating this page into {lang}">join the discussion</a>."
> ```
> 
> Then, we can add the chapter in English to every translation folder with a Front-Matter param containing the Id of the the translation issue. If the param is set, the abovementioned piece of text is displayed, the html tag displays an "en" `lang` attribute, and the page is not indexed. If the param is not set, the page is fully translated.

So you're going with @rviscomi 's suggestion of having it in English but with a note at the top? What do you think of the idea of machine translating the chapters in the meantime? Don't think that's a good idea?

Also this could cause problems with sitemap and `xhreflang` links. Guessing while we have `no-index` we should include the sitemap and `xhreflang` info in that. Again machine translating might at least give the "real language" (even if a poor translation) so avoid those complications and mean the language is really live for all chapters rather than half live.

I do worry about having single chapters translated for a long time. E.g. a Dutch SEO expert takes it upon themselves to translate the SEO chapter to Dutch but doesn't have the interest in translating all the other chapters. Do we only launch the SEO chapter and the other chapters aren't avaialble to view in Dutch (more complex routing logic)? Do we launch this with potentially English versions for the other 19 chapters under the `nl` URL for a long time (which introduces other problems)?  Do we machine translate the others so at least ahve a Dutch version of them (and, if so, what's the effort in that)?

> In the language selector, indeed, we can add an "add a translation" option, then redirect people to… [this research](https://github.com/HTTPArchive/almanac.httparchive.org/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+%28translate+OR+translation%29)?

We have a `Translation` label btw so can use this: https://github.com/HTTPArchive/almanac.httparchive.org/labels/Translation which also avoids any false positives (assuming the label is used correctly but I'm pretty good at ensuring that).

I do like the idea of making it easy for people to direct people who want to start a new translation. However, on the flip side, it does risk adding more languages and complexity that we then don't make progress on (e.g. Arabic, Portuguese, Russian...etc.). that's fine if no code is commited (Portuguese, Russian) as just an issue, but less fine once we add code that needs to be maintained (Arabic). Though again once #676 lands the maintenance is a lot less.

> The only negative side of this approach is that the URLs would not be located. I do not realize the importance of the problem, but I think that, at worst, if we later realize that it is inconvenient, we will be able to implement a fix without invalidating the work previously done.

Sorry, I'm not sure what you mean by this?

**Comment by bazzadp** [comment]
Made some good progress on this in #684 

One thing I haven't done that maybe should consider is checking if chapter exists and if not flagging in different way in ToC and in next/previous chapter widgets so people realise they will only get English version.

**Comment by bazzadp** [comment]
> One thing I haven't done that maybe should consider is checking if chapter exists and if not flagging in different way in ToC and in next/previous chapter widgets so people realise they will only get English version.

Added this to #684 . Could probably do with some design tips but it works.

**Comment by bazzadp** [comment]
OK I think we're in a good place to release French and Spanish. I'd love to get Japanese in there in the first tranch as they have a load of chapters ready but just [need two of the base templates translated for that to be possible](https://github.com/HTTPArchive/almanac.httparchive.org/issues/586#issuecomment-600262540).

Other than that you happy to release @rviscomi and any particular day or time you want to launch this?

**Comment by rviscomi** [comment]
Thanks Barry! Launching as soon as the translations are ready SGTM, no particular day/time. Let's coordinate in Slack when things go live so I can help promote from @HTTPArchive on Twitter.

How much of a blocker are the base templates for the Japanese translations? Is it a deal-breaker to have part of the site temporarily untranslated? (That might be a nice motivator to draw in translation assistance)

**Comment by bazzadp** [comment]
> How much of a blocker are the base templates for the Japanese translations? Is it a deal-breaker to have part of the site temporarily untranslated? (That might be a nice motivator to draw in translation assistance)

Well they need to exist before the chapters will render so can’t turn on Japanese without them.

We could just upload the English versions to the `ja` folder I suppose. They aren’t really templates anymore - just a list of content blocks that are slotted into the actual pages (see [base.html](https://github.com/HTTPArchive/almanac.httparchive.org/blob/master/src/templates/en/2019/base.html) and [base_chapter.html](https://github.com/HTTPArchive/almanac.httparchive.org/blob/master/src/templates/en/2019/base_chapter.html) to see what I mean). So things like “Index”, “Next Chapter”, “Previous Chapter”...etc. Plus all the content to create the translated header and footer. So not super critical I guess and could live with English for now. A shame though cause they really are quite short to translate and do think it would mean the whole page looks translated.

The again visualisations (#558) and Logo (#567) won’t be translated so maybe I am just being too perfectionist here and we should go with English?

**Comment by rviscomi** [comment]
Yeah I'd default to English for anything untranslated just to get it out the door as quickly as possible. I'm sure the lack of translation will bother people just enough to do it themselves :)

That reminds me, another nice feature would be a "suggest an edit" UI that enables visitors to step directly into a GitHub edit flow.

**Comment by bazzadp** [comment]
> Yeah I'd default to English for anything untranslated just to get it out the door as quickly as possible. I'm sure the lack of translation will bother people just enough to do it themselves :)

Submitted #690. Had a few of the important terms translated anyway so doesn't look too bad.

> That reminds me, another nice feature would be a "suggest an edit" UI that enables visitors to step directly into a GitHub edit flow.

Kind of fits in with #518 for other info we were going to add to chapters. Though those less familiar with Git may prefer to raise Issues than go straight into Editing.

**Comment by bazzadp** [comment]
Closing this as [we're live with French, Spanish and Japanese](https://twitter.com/HTTPArchive/status/1241093228562132993?s=20)!
